File: haskell/2015/Day2.hs
module Day2 where

import Data.List
import Data.List.Split

type Dimensions = [Int]

determineAreaNeeded :: [Int] -> Int
determineAreaNeeded [l, w, h] = sum sides + minimum sides
  where
    sides = [l * w, l * w, l * h, l * h, w * h, w * h]

determineLengthNeeded :: [Int] -> Int
determineLengthNeeded xs = product xs + length' xs
  where
    length' = sum . concatMap (\x -> [x, x]) . init . sort

main' :: IO ()
main' = do
  contents <- map (map read . splitOn "x") . lines <$> readFile "2015/inputs/Day2/input.txt" :: IO [[Int]]
  -- part 1
  print $ sum $ map determineAreaNeeded contents
  -- part 2
  print $ sum $ map determineLengthNeeded contents

File: elixir/2015/lib/day2.ex
defmodule Day2 do
  @spec determineAreaNeeded([integer()]) :: integer()
  def determineAreaNeeded([l, w, h]) do
    sides = [l * w, l * w, l * h, l * h, w * h, w * h]
    Enum.sum(sides) + Enum.min(sides)
  end

  @spec determineLengthNeeded([integer()]) :: integer()
  def determineLengthNeeded(list) do
    length =
      list |> Enum.sort() |> Enum.drop(-1) |> Enum.flat_map(fn x -> [x, x] end) |> Enum.sum()

    Enum.product(list) + length
  end

  def main() do
    contents =
      File.read!("../../inputs/2015/Day2/input.txt")
      |> String.split("\n")
      # drop last empty line
      |> Enum.drop(-1)
      |> Enum.map(fn s -> s |> String.split("x") |> Enum.map(&String.to_integer(&1)) end)

    part1 = Enum.map(contents, &Day2.determineAreaNeeded(&1)) |> Enum.sum()
    IO.inspect(part1)

    part2 = Enum.map(contents, &Day2.determineLengthNeeded(&1)) |> Enum.sum()
    IO.inspect(part2)
  end
end
